main.tf

```diff
@@ -1,33 +1,207 @@
+
+  public_network_route_enabled = local.public_enabled && var.public_network_route_enabled
+  public_network_table_enabled = local.public_network_route_enabled && (length(var.public_route_table_id) == 0 || local.public_dns64_enabled)
+  public_network_table_count   = local.public_network_table_enabled ? (local.public_dns64_enabled ? local.subnet_az_count : 1) : 0
+  public_network_table_ids     = local.public_network_table_enabled ? aws_route_table.public.*.id : var.public_route_table_id
+
+  private_network_route_enabled = local.private_enabled && (local.nat_enabled || local.ipv6_egress_only_configured) && var.private_network_route_enabled
+  private_network_table_enabled = local.private_network_route_enabled
+  private_network_table_count   = local.private_network_table_enabled ? local.subnet_az_count : 0
+  private_network_table_ids     = local.private_network_table_enabled ? aws_route_table.private.*.id : []
+
+  # public and private network ACLs
+  # Support deprecated var.public_network_acl_id
+  public_open_network_acl_enabled = local.public_enabled && (length(var.public_network_acl_id) > 0 ? false : var.public_open_network_acl_enabled)
+  # Support deprecated var.private_network_acl_id
+  private_open_network_acl_enabled = local.private_enabled && (length(var.private_network_acl_id) > 0 ? false : var.private_open_network_acl_enabled)
+
+  # A NAT device is needed to NAT from private IPv4 to public IPv4 or to perform NAT64 for IPv6,
+  # but since it must be placed in a public subnet, we consider it not required if we are not creating public subnets.
+  nat_required        = local.public_enabled && (local.private4_enabled || local.public_dns64_enabled)
+  nat_gateway_enabled = local.nat_required && var.nat_gateway_enabled
+  # An AWS NAT instance does not perform NAT64, but we allow for one anyway because
+  # the user can supply their own NAT instance AMI that does support it.
+  nat_instance_enabled = local.nat_required && var.nat_instance_enabled
+  nat_enabled          = local.nat_gateway_enabled || local.nat_instance_enabled
+  need_nat_eips        = local.nat_enabled && length(var.nat_elastic_ips) == 0
+  need_nat_eip_data    = local.nat_enabled && length(var.nat_elastic_ips) > 0
+  nat_eip_allocations  = local.nat_enabled ? (local.need_nat_eips ? aws_eip.default.*.id : data.aws_eip.nat.*.id) : []
+
+  need_nat_ami_id     = local.nat_instance_enabled && length(var.nat_instance_ami_id) == 0
+  nat_instance_ami_id = local.need_nat_ami_id ? data.aws_ami.nat_instance[0].id : try(var.nat_instance_ami_id[0], "")
+
 }
 
-data "aws_availability_zones" "available" {
-  count = local.enabled_count
+data "aws_availability_zones" "default" {
+  count = local.enabled ? 1 : 0
+
+  # Filter out Local Zones. See https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/availability_zones#by-filter
+  filter {
+    name   = "opt-in-status"
+    values = ["opt-in-not-required"]
+  }
```

Comment:
No, because we want above all else to assign consistent CIDRs to each subnet, and part of that means reserving CIDRs for each AZ, whether it is available or not. We also do not want to avoid creating a subnet in a region just because, by coincidence, it is currently "impaired" at the time you are deploying the subnets. We generally recommend specifying AZs explicitly, either by name or ID, rather than auto-assigning AZs, for these or other reasons.

Reply:
Okay, when explained like that it makes sense to include all zones regardless of `state`.